lib/Frontend/CompilerInvocation.cpp

```diff
@@ -83,6 +84,159 @@ SourceFileKind CompilerInvocation::getSourceFileKind() const {
+  };
+
+public:
+  ArgsToFrontendInputsConverter(DiagnosticEngine &Diags, const ArgList &Args,
+                                FrontendInputs &Inputs)
+      : Diags(Diags), Args(Args), Inputs(Inputs) {}
+
+  bool convert() {
+    if (enforceFilelistExclusion())
+      return true;
+    const FilesInfo info = getFiles();
+    if (info.hadError)
+      return true;
+    setInputFilesAndIndices(info);
+    return setPrimaryFiles(info);
+  }
+
+private:
+  bool enforceFilelistExclusion() {
+    if (Args.hasArg(options::OPT_INPUT) && Args.hasArg(options::OPT_filelist)) {
+      Diags.diagnose(SourceLoc(),
+                     diag::error_cannot_have_input_files_with_file_list);
+      return true;
+    }
+    return false;
+  }
+
+  FilesInfo getFiles() {
+    // Even if filelist is present must do this to read primary file arguments.
+    FilesInfo info = getFilesFromCommandLine();
+    if (info.hadError)
+      return info;
+    addFilesFromFilelistTo(info);
+    return info;
+  }
+
+  FilesInfo getFilesFromCommandLine() {
+    std::vector<std::pair<StringRef, FileKind>> files;
+    for (const Arg *A :
+         Args.filtered(options::OPT_INPUT, options::OPT_primary_file)) {
+      FileKind fileType;
+      if (A->getOption().matches(options::OPT_INPUT)) {
+        fileType = FileKind::Secondary;
+      } else if (A->getOption().matches(options::OPT_primary_file)) {
+        fileType = FileKind::Primary;
+      } else {
+        llvm_unreachable("Unknown input-related argument!");
+      }
+      files.push_back(std::make_pair(A->getValue(), fileType));
+    }
+    return FilesInfo::fromCommandLine(std::move(files));
+  }
+
+  void addFilesFromFilelistTo(FilesInfo &info) {
+    const llvm::opt::Arg *filelistArg = Args.getLastArg(options::OPT_filelist);
+    if (filelistArg == nullptr)
+      return;
+    info.filelistPath = filelistArg->getValue();
+    llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> filelistBufferOrError =
+        llvm::MemoryBuffer::getFile(info.filelistPath);
+    if (!filelistBufferOrError) {
+      Diags.diagnose(SourceLoc(), diag::cannot_open_file, info.filelistPath,
+                     filelistBufferOrError.getError().message());
+      info.hadError = true;
+      return;
+    }
+    // Keep buffer alive because code passes around StringRefs.
+    FilelistBuffer = std::move(*filelistBufferOrError);
+    std::vector<StringRef> inputFilesFromFilelist(
+        llvm::line_iterator(*FilelistBuffer), {});
+    for (auto file : inputFilesFromFilelist) {
+      info.files.push_back(std::make_pair(file, FileKind::Secondary));
+    }
+  }
+
+  void setInputFilesAndIndices(const FilesInfo &info) {
+    for (std::pair<StringRef, FileKind> p : info.files) {
+      if (p.second == FileKind::Secondary ||
+          info.mustAddPrimariesToAllFiles()) {
+        unsigned index = Inputs.inputFilenameCount();
+        auto file = p.first;
+        Inputs.addInputFilename(file);
```

Comment:
…ah. You just have primary files showing up twice in the list in filelist cases. I don't really think that's better than what we had before with two separate code paths.